compiler/rustc_metadata/src/locator.rs

```diff
@@ -783,14 +789,67 @@ fn get_metadata_section<'p>(
     filename: &'p Path,
     loader: &dyn MetadataLoader,
     cfg_version: &'static str,
+    crate_kind: ExternCrateKind,
+    crate_name: Option<Symbol>,
 ) -> Result<MetadataBlob, MetadataError<'p>> {
     if !filename.exists() {
         return Err(MetadataError::NotPresent(filename));
     }
+
+    // FIXME: Allow it for statically linked dependencies as well.
+    if crate_kind == ExternCrateKind::Stable && flavor != CrateFlavor::Dylib {
+        return Err(MetadataError::LoadFailure(
+            "`extern dyn` annotation is only avaible for dynamic dependencies".to_string(),
+        ));
+    }
+
     let raw_bytes = match flavor {
         CrateFlavor::Rlib => {
             loader.get_rlib_metadata(target, filename).map_err(MetadataError::LoadFailure)?
         }
+        CrateFlavor::Dylib if crate_kind == ExternCrateKind::Stable => {
```

Comment:
I think we should use an entirely separate crate flavor for dylibs with a stable interface.